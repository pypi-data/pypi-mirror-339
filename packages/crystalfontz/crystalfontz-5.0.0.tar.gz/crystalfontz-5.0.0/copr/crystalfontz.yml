apiVersion: coprctl/v1alpha1
kind: package-scm
metadata:
  name: crystalfontz
spec:
  auto_rebuild: true
  ownername: jfhbrook
  projectname: joshiverse
  clone-url: https://github.com/jfhbrook/crystalfontz
  commit: null
  spec: crystalfontz.spec
  type: git
  method: tito
