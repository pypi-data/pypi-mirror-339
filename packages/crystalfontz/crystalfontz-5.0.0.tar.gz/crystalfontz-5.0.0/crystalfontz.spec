Name: crystalfontz
Version: 5.0.0
Release: 1
License: MPL-2.0
Summary: Serial client and Linux service for Crystalfontz LCD displays

URL: https://github.com/jfhbrook/crystalfontz
Source0: %{name}-%{version}.tar.gz
BuildArch: noarch

Requires: python-crystalfontz
Requires: python-sdbus

%description


%prep
%autosetup


%build
tar -xzf %{SOURCE0}


%install
mkdir -p %{buildroot}%{_prefix}/lib/systemd/system
mkdir -p %{buildroot}%{_bindir}
install -p -D -m 0644 systemd/crystalfontz.service %{buildroot}%{_prefix}/lib/systemd/system/crystalfontz.service
install -p -D -m 0644 dbus/org.jfhbrook.crystalfontz.conf %{buildroot}%{_prefix}/share/dbus-1/system.d/org.jfhbrook.crystalfontz.conf
install -p -m 755 bin/crystalfontz-dbus %{buildroot}%{_bindir}/crystalfontz

%check


%files
%{_prefix}/lib/systemd/system/crystalfontz.service
%{_prefix}/share/dbus-1/system.d/org.jfhbrook.crystalfontz.conf
%{_bindir}/crystalfontz

%changelog
* Tue Apr 08 2025 Josh Holbrook <josh.holbrook@gmail.com> 5.0.0-1
- API Changes:
  - **NEW:** `Receiver` class
    - A subclass of `asyncio.Queue`
  - `Client`:
    - **BREAKING:** `subscribe` and `unsubscribe` methods use new `Receiver` class
    - Emit unmatched exceptions on expecting receivers instead of resolving `client.closed`
    - `detect_baud_rate` exposes `timeout` and `retry_times` arguments
    - Document `detect_baud_rate`
  - **BREAKING:** `ClientProtocol`/`EffectClient`
    - `crystalfontz.protocol.ClientProtocol` type has been replaced by `crystalfontz.effects.EffectClient`
    - `EffectClient` enforces a smaller API than `ClientProtocol` did previously
  - **BREAKING:** `Response`:
    - `Response.from_bytes` accepts bytes as from packets, rather than `__init__`
    - `Response.__init__` accepts properties as arguments
  - **BREAKING:** `SpecialCharacter` API:
    - Rename `as_bytes` method to `to_bytes`
    - Store pixels as `List[List[bool]]` instead of `List[List[int]]`
  - `KeyState` includes `keypress: KeyPress` attribute
- CLI changes:
  - **BREAKING:** CLI now invoked through `python3 -m crystalfontz`
    - Optional unpackaged `crystalfontz` entry point at `./bin/crystalfontz`
  - Use `configurence` library
  - Respect `CRYSTALFONTZ_CONFIG_FILE` environment variable
  - Improve error reporting for timeouts
  - Client now respects CLI settings
    - Cause was the `Client` constructor being called twice
  - Additional commands accept bytes as arguments
    - `python3 -m crystalfontz line 1`
    - `python3 -m crystalfontz line 2`
    - `python3 -m crystalfontz send`
  - Help for `python3 -m crystalfontz listen` `--for` option
- **NEW:** Dbus support:
  - `crystalfontz.dbus.DbusInterface` dbus Interface class, implementing most commands
  - `crystalfontz.dbus.DbusClient` dbus client class
  - `crystalfontz.dbus.domain` API for mapping domain objects to dbus types
  - `python3 -m crystalfontz.dbus.service` dbus service CLI
    - Optional unpackaged `crystalfontz.dbus.service` entry point at `./bin/crystalfontz-service`
  - `python3 -m crystalfontz.dbus.client` dbus client CLI
    - Optional unpackaged `crystalfontz.dbus.client` entry point at `./bin/crystalfontz-dbus`
- Packaging and Releases:
  - `python-crystalfontz` COPR package spec
    - This has been moved from <https://github.com/jfhbrook/public>
  - **NEW:** `crystalfontz` COPR package
    - Depends on `python-crystalfontz` COPR package
    - Includes systemd unit for `python3 -m crystalfontz.dbus.service`
    - **BREAKING:** Includes shim bin `crystalfontz` -> `python3 -m crystalfontz.dbus.client`
  - Tito based release tagging
  - GitHub release
  - Improved PyPI classifiers
  - **BREAKING:** Release under MPL-2.0 license
- Integration tests:
  - Use `python-gak` plugin
  - Use snapshots
  - Leverage a config file at `./tests/fixtures/crystalfontz.yaml`
    - Can be overridden with `CRYSTALFONTZ_CONFIG_FILE` environment variable


