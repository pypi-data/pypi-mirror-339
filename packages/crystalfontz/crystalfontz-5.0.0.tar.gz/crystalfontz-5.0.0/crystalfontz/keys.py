from dataclasses import asdict, dataclass
from enum import Enum
from typing import List, Literal, Self, Tuple, Type

KeyPress = (
    Literal[0x01]
    | Literal[0x02]
    | Literal[0x04]
    | Literal[0x08]
    | Literal[0x10]
    | Literal[0x20]
)

KP_UP: KeyPress = 0x01
KP_ENTER: KeyPress = 0x02
KP_EXIT: KeyPress = 0x04
KP_LEFT: KeyPress = 0x08
KP_RIGHT: KeyPress = 0x10
KP_DOWN: KeyPress = 0x20


def keypress_repr(keypress: KeyPress) -> str:
    return {
        KP_UP: "KP_UP",
        KP_ENTER: "KP_ENTER",
        KP_EXIT: "KP_EXIT",
        KP_LEFT: "KP_LEFT",
        KP_RIGHT: "KP_RIGHT",
        KP_DOWN: "KP_DOWN",
    }[keypress]


@dataclass
class KeyState:
    """
    A key's state.

    Attributes:
        pressed (bool): When True, the key is currently pressed.
        pressed_since (bool): When True, the key has been pressed since the last poll.
        released_since (bool): When True, the key has been released since the last
                               poll.
    """

    keypress: KeyPress
    pressed: bool
    pressed_since: bool
    released_since: bool

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes, keypress: KeyPress) -> Self:
        pressed = state[0]
        pressed_since = state[1]
        released_since = state[2]

        return cls(
            keypress=keypress,
            pressed=bool(pressed & keypress),
            pressed_since=bool(pressed_since & keypress),
            released_since=bool(released_since & keypress),
        )

    def to_bytes(self: Self) -> Tuple[int, int, int]:
        pressed = self.keypress if self.pressed else 0x00
        pressed_since = self.keypress if self.pressed_since else 0x00
        released_since = self.keypress if self.released_since else 0x00

        return (pressed, pressed_since, released_since)


@dataclass
class KeyStates:
    """
    The state of all keys.

    Attributes:
        up: The state of the "up" key.
        enter: The state of the "enter" key.
        exit: The state of the "exit" key.
        left: The state of the "left" key.
        right: The state of the "right" key.
        down: The state of the "down" key.
    """

    up: KeyState
    enter: KeyState
    exit: KeyState
    left: KeyState
    right: KeyState
    down: KeyState

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes) -> Self:
        return cls(
            up=KeyState.from_bytes(state, KP_UP),
            enter=KeyState.from_bytes(state, KP_ENTER),
            exit=KeyState.from_bytes(state, KP_EXIT),
            left=KeyState.from_bytes(state, KP_LEFT),
            right=KeyState.from_bytes(state, KP_RIGHT),
            down=KeyState.from_bytes(state, KP_DOWN),
        )

    def to_bytes(self: Self) -> bytes:
        pressed = 0x00
        pressed_since = 0x00
        released_since = 0x00

        for state in [
            self.up,
            self.enter,
            self.exit,
            self.left,
            self.right,
            self.down,
        ]:
            p, p_s, r_s = state.to_bytes()
            pressed = pressed ^ p
            pressed_since = pressed_since ^ p_s
            released_since = released_since ^ r_s

        return bytes([pressed, pressed_since, released_since])

    def __repr__(self: Self) -> str:
        repr_ = ""
        for name, state in asdict(self).items():
            st = ", ".join(
                [
                    (
                        f"{n}={'yes' if s else 'no'}"
                        if isinstance(s, bool)
                        else f"{n}={keypress_repr(s)}"
                    )
                    for n, s in state.items()
                ]
            )
            repr_ += f"{name}: {st}\n"
        return repr_[0:-1]


class KeyActivity(Enum):
    """
    A key activity. This is either a "press" event or a "release" event for a
    given key.
    """

    KEY_UP_PRESS = 1
    KEY_DOWN_PRESS = 2
    KEY_LEFT_PRESS = 3
    KEY_RIGHT_PRESS = 4
    KEY_ENTER_PRESS = 5
    KEY_EXIT_PRESS = 6
    KEY_UP_RELEASE = 7
    KEY_DOWN_RELEASE = 8
    KEY_LEFT_RELEASE = 9
    KEY_RIGHT_RELEASE = 10
    KEY_ENTER_RELEASE = 11
    KEY_EXIT_RELEASE = 12

    @classmethod
    def from_byte(cls: Type[Self], activity: int) -> "KeyActivity":
        return KEY_ACTIVITIES[activity - 1]

    @classmethod
    def from_bytes(cls: Type[Self], activity: bytes) -> "KeyActivity":
        return cls.from_byte(activity[0])

    def to_byte(self: Self) -> int:
        return self.value


KEY_ACTIVITIES: List[KeyActivity] = [
    KeyActivity.KEY_UP_PRESS,
    KeyActivity.KEY_DOWN_PRESS,
    KeyActivity.KEY_LEFT_PRESS,
    KeyActivity.KEY_RIGHT_PRESS,
    KeyActivity.KEY_ENTER_PRESS,
    KeyActivity.KEY_EXIT_PRESS,
    KeyActivity.KEY_UP_RELEASE,
    KeyActivity.KEY_DOWN_RELEASE,
    KeyActivity.KEY_LEFT_RELEASE,
    KeyActivity.KEY_RIGHT_RELEASE,
    KeyActivity.KEY_ENTER_RELEASE,
    KeyActivity.KEY_EXIT_RELEASE,
]
