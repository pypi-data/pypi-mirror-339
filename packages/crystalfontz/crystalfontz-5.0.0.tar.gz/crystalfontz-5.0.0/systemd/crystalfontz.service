[Unit]
Description=Crystalfontz LCD
Requires=dbus.socket
After=dbus.socket

[Service]
Type=dbus
BusName=org.jfhbrook.crystalfontz
ExecStart=/usr/bin/python3 -m crystalfontz.dbus.service --system
Restart=on-failure

[Install]
WantedBy=multi-user.target
