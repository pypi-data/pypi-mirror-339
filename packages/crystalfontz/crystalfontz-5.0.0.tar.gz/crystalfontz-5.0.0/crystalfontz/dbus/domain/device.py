from typing import ClassVar, Tuple

from crystalfontz.dbus.domain.base import (
    BytesM,
    ModelM,
    ModelT,
    RevisionM,
    RevisionT,
    struct,
    t,
    Uint16M,
    Uint16T,
)
from crystalfontz.device import Device, DeviceStatus, lookup_device

WidthT = Uint16T


class WidthM(Uint16M):
    pass


HeightT = Uint16T


class HeightM(Uint16M):
    pass


DeviceT = Tuple[ModelT, RevisionT, RevisionT, WidthT, HeightT, WidthT, HeightT, Uint16T]


class DeviceM:
    t: ClassVar[str] = struct(
        ModelM, RevisionM, RevisionM, WidthM, HeightM, WidthM, HeightM, Uint16M
    )

    @staticmethod
    def pack(device: Device) -> DeviceT:
        return (
            device.model,
            device.hardware_rev,
            device.firmware_rev,
            device.lines,
            device.columns,
            device.character_width,
            device.character_height,
            device.n_temperature_sensors,
        )

    @staticmethod
    def unpack(device: DeviceT) -> Device:
        model = device[0]
        hardware_rev = device[1]
        firmware_rev = device[2]

        return lookup_device(model, hardware_rev, firmware_rev)


DeviceStatusT = Tuple[ModelT, RevisionT, RevisionT, bytes]


class DeviceStatusM:
    t: ClassVar[str] = t(ModelM, RevisionM, RevisionM, BytesM)

    @staticmethod
    def pack(status: DeviceStatus, device: Device) -> DeviceStatusT:
        return (
            device.model,
            device.hardware_rev,
            device.firmware_rev,
            status.to_bytes(device),
        )

    @staticmethod
    def unpack(status: DeviceStatusT) -> DeviceStatus:
        model, hardware_rev, firmware_rev, data = status
        device = lookup_device(model, hardware_rev, firmware_rev)
        return device.status(data)
