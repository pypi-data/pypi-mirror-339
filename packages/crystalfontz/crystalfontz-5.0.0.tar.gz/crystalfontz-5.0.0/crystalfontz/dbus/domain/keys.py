from typing import ClassVar, Set, Tuple

from crystalfontz.dbus.domain.base import ByteM, OptFloatM, OptFloatT, struct
from crystalfontz.keys import (
    KeyActivity,
    KeyPress,
    KeyState,
    KeyStates,
    KP_DOWN,
    KP_ENTER,
    KP_EXIT,
    KP_LEFT,
    KP_RIGHT,
    KP_UP,
)

KeyPressT = int

KEYPRESSES: Set[KeyPress] = {KP_UP, KP_ENTER, KP_EXIT, KP_LEFT, KP_RIGHT, KP_DOWN}


class KeyPressM(ByteM):
    t: ClassVar[str] = ByteM.t

    @staticmethod
    def unpack(keypress: int) -> KeyPress:
        if keypress not in KEYPRESSES:
            raise ValueError(f"{keypress} is not a valid key press")
        return keypress


KeyStateT = Tuple[bool, bool, bool]
KeyStatesT = Tuple[
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
]


class KeyStateM:
    t: ClassVar[str] = struct("bbb")

    @staticmethod
    def pack(state: KeyState) -> KeyStateT:
        return (state.pressed, state.pressed_since, state.released_since)

    @staticmethod
    def unpack(state: KeyStateT, keypress: KeyPress) -> KeyState:
        pressed, pressed_since, released_since = state
        return KeyState(
            keypress=keypress,
            pressed=pressed,
            pressed_since=pressed_since,
            released_since=released_since,
        )


class KeyStatesM:
    t: ClassVar[str] = struct(KeyStateM.t) * 6

    @staticmethod
    def pack(states: KeyStates) -> KeyStatesT:
        return (
            KeyStateM.pack(states.up),
            KeyStateM.pack(states.enter),
            KeyStateM.pack(states.exit),
            KeyStateM.pack(states.left),
            KeyStateM.pack(states.right),
            KeyStateM.pack(states.down),
        )

    @staticmethod
    def unpack(states: KeyStatesT) -> KeyStates:
        up, enter, exit, left, right, down = states
        return KeyStates(
            up=KeyStateM.unpack(up, KP_UP),
            enter=KeyStateM.unpack(enter, KP_ENTER),
            exit=KeyStateM.unpack(exit, KP_EXIT),
            left=KeyStateM.unpack(left, KP_LEFT),
            right=KeyStateM.unpack(right, KP_RIGHT),
            down=KeyStateM.unpack(down, KP_DOWN),
        )


KeypadBrightnessT = OptFloatT


class KeypadBrightnessM(OptFloatM):
    """
    Map `Optional[float]` to and from `KeypadBrightnessT` (`float`).

    `KeypadBrightnessM` is an alias for `OptFloatM`.
    """

    t: ClassVar[str] = OptFloatM.t


KeyActivityT = int


class KeyActivityM:
    """
    Map `KeyActivity` to and from `KeyActivityT` (`int`).
    """

    t: ClassVar[str] = ByteM.t

    @staticmethod
    def pack(activity: KeyActivity) -> KeyActivityT:
        return activity.to_byte()

    @staticmethod
    def unpack(activity: KeyActivityT) -> KeyActivity:
        return KeyActivity.from_byte(activity)
