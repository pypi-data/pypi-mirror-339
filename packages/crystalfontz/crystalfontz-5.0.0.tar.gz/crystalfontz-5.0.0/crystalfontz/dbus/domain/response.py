from typing import ClassVar, Tuple

from crystalfontz.dbus.domain.base import (
    AddressM,
    AddressT,
    ByteM,
    BytesM,
    ByteT,
    IndexM,
    IndexT,
    ModelM,
    ModelT,
    RevisionM,
    RevisionT,
    struct,
    t,
    Uint16M,
    Uint16T,
)
from crystalfontz.dbus.domain.gpio import (
    GpioSettingsM,
    GpioSettingsT,
    GpioStateM,
    GpioStateT,
)
from crystalfontz.dbus.domain.keys import (
    KeyActivityM,
    KeyActivityT,
    KeyStatesM,
    KeyStatesT,
)
from crystalfontz.response import (
    DowDeviceInformation,
    DowTransactionResult,
    GpioRead,
    KeyActivityReport,
    KeypadPolled,
    LcdMemory,
    Pong,
    TemperatureReport,
    UserFlashAreaRead,
    Versions,
)

PongT = bytes


class PongM:
    """
    Map `Pong` to and from `PongT` (`bytes`).
    """

    t: ClassVar[str] = BytesM.t

    @staticmethod
    def pack(pong: Pong) -> PongT:
        """
        Pack `Pong` to `PongT`.
        """

        return pong.response

    @staticmethod
    def unpack(pong: PongT) -> Pong:
        """
        Unpack `PongT` to `Pong`.
        """

        return Pong(pong)


VersionsT = Tuple[ModelT, RevisionT, RevisionT]


class VersionsM:
    """
    Map `Versions` to and from `VersionsT` (`Tuple[str, str, str]`).
    """

    t: ClassVar[str] = t(ModelM, RevisionM, RevisionM)

    @staticmethod
    def pack(versions: Versions) -> VersionsT:
        return (versions.model, versions.hardware_rev, versions.firmware_rev)

    @staticmethod
    def unpack(versions: VersionsT) -> Versions:
        """
        Unpack `VersionsT` to `Versions`.
        """

        return Versions(*versions)


UserFlashAreaReadT = bytes


class UserFlashAreaReadM:
    """
    Map `UserFlashAreaRead` to and from `UserFlashAreaReadT` (`bytes`).
    """

    t: ClassVar[str] = BytesM.t

    @staticmethod
    def pack(res: UserFlashAreaRead) -> UserFlashAreaReadT:
        return res.data

    @staticmethod
    def unpack(res: UserFlashAreaReadT) -> UserFlashAreaRead:
        """
        Unpack `UserFlashAreaReadT` to `UserFlashAreaRead`.
        """

        return UserFlashAreaRead(res)


LcdMemoryT = Tuple[AddressT, bytes]


class LcdMemoryM:
    """
    Map `LcdMemory` to and from `LcdMemoryT` (`Tuple[int, bytes]`).
    """

    t: ClassVar[str] = t(AddressM, BytesM)

    @staticmethod
    def pack(memory: LcdMemory) -> LcdMemoryT:
        return (memory.address, memory.data)

    @staticmethod
    def unpack(obj: LcdMemoryT) -> LcdMemory:
        """
        Unpack `LcdMemoryT` to `LcdMemory`.
        """

        address, buff = obj
        return LcdMemory(address, buff)


DowDeviceInformationT = Tuple[IndexT, int]


class DowDeviceInformationM:
    """
    Map `DowDeviceInformation` to and from `DowDeviceInformationT`
    (`Tuple[int, int]`).
    """

    t: ClassVar[str] = t(IndexM, "t")

    @staticmethod
    def pack(info: DowDeviceInformation) -> DowDeviceInformationT:
        return (info.index, info.rom_id)

    @staticmethod
    def unpack(info: DowDeviceInformationT) -> DowDeviceInformation:
        """
        Unpack `DowDeviceInformation` to `DowDeviceInformationT`.
        """

        index, rom_id = info
        return DowDeviceInformation(index, rom_id)


DowTransactionResultT = Tuple[IndexT, bytes, Uint16T]


class DowTransactionResultM:
    """
    Map `DowTransactionResult` to and from `DowTransactionResultT`
    (`Tuple[int, bytes, int]`).
    """

    t: ClassVar[str] = t(IndexM, BytesM, Uint16M)

    @staticmethod
    def pack(res: DowTransactionResult) -> DowTransactionResultT:
        return (res.index, res.data, res.crc)

    @staticmethod
    def unpack(res: DowTransactionResultT) -> DowTransactionResult:
        """
        Unpack `DowTransactionResultT` to `DowTransactionResult`.
        """

        index, data, crc = res
        return DowTransactionResult(index, data, crc)


KeypadPolledT = KeyStatesT


class KeypadPolledM:
    """
    Map `KeypadPolled` to and from `KeypadPolledT`
    (`Tuple[KeyStateT, KeyStateT, KeyStateT, KeyStateT, KeyStateT, KeyStateT]`,
    where `KeyStateT = Tuple[bool, bool, bool]`).
    """

    t: ClassVar[str] = KeyStatesM.t

    @staticmethod
    def pack(polled: KeypadPolled) -> KeyStatesT:
        return KeyStatesM.pack(polled.states)

    @staticmethod
    def unpack(polled: KeypadPolledT) -> KeypadPolled:
        """
        Unpack `KeypadPolledT` to `KeypadPolled`.
        """

        return KeypadPolled(KeyStatesM.unpack(polled))


GpioReadT = Tuple[IndexT, GpioStateT, ByteT, GpioSettingsT]


class GpioReadM:
    """
    Map `GpioRead` to and from `GpioReadT`
    (`Tuple[int, Tuple[bool, bool, bool], int, int]`).
    """

    t: ClassVar[str] = t(IndexM, struct(GpioStateM), ByteM, GpioSettingsM)

    @staticmethod
    def pack(gpio_read: GpioRead) -> GpioReadT:
        return (
            gpio_read.index,
            GpioStateM.pack(gpio_read.state),
            gpio_read.requested_level,
            GpioSettingsM.pack(gpio_read.settings),
        )

    @staticmethod
    def unpack(gpio_read: GpioReadT) -> GpioRead:
        """
        Unpack `GpioReadT` to `GpioRead`.
        """

        index, state, requested_level, settings = gpio_read
        return GpioRead(
            index=index,
            state=GpioStateM.unpack(state),
            requested_level=requested_level,
            settings=GpioSettingsM.unpack(settings),
        )


KeyActivityReportT = KeyActivityT


class KeyActivityReportM:
    """
    Map `KeyActivityReport` to and from `KeyActivityReportT`
    """

    t: ClassVar[str] = KeyActivityM.t

    @staticmethod
    def pack(report: KeyActivityReport) -> KeyActivityReportT:
        return KeyActivityM.pack(report.activity)

    @staticmethod
    def unpack(report: KeyActivityReportT) -> KeyActivityReport:
        return KeyActivityReport(KeyActivityM.unpack(report))


TemperatureReportT = Tuple[IndexT, float, float]


class TemperatureReportM:
    """
    Map `TemperatureReport` to and from `KeyActivityReportT`
    """

    t: ClassVar[str] = t(IndexM, "dd")

    @staticmethod
    def pack(report: TemperatureReport) -> TemperatureReportT:
        return (report.index, report.celsius, report.fahrenheit)

    @staticmethod
    def unpack(report: TemperatureReportT) -> TemperatureReport:
        index, celsius, fahrenheit = report
        return TemperatureReport(index, celsius, fahrenheit)
