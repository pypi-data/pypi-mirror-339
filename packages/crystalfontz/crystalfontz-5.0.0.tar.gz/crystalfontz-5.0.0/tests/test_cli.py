import json
from typing import Any

import pytest

from crystalfontz.atx import AtxPowerSwitchFunction, AtxPowerSwitchFunctionalitySettings
from crystalfontz.cli import as_json, parse_bytes
from crystalfontz.device import CFA533Status
from crystalfontz.keys import (
    KeyState,
    KeyStates,
    KP_DOWN,
    KP_ENTER,
    KP_EXIT,
    KP_LEFT,
    KP_RIGHT,
    KP_UP,
)
from crystalfontz.response import (
    DowDeviceInformation,
    DowTransactionResult,
    GpioRead,
    KeypadPolled,
    LcdMemory,
    Versions,
)


@pytest.mark.parametrize(
    "text,buffer",
    [
        ("hello world", b"hello world"),
        ("\\\\", b"\\"),
        ("\\a", b"\a"),
        ("\\o333", (0o333).to_bytes(1, "big")),
        ("\\o22", (0o22).to_bytes(1, "big")),
        ("\\xff", b"\xff"),
        ("\\", b"\\"),
        ("\\s", b"\\s"),
        ("\\xf", b"\\xf"),
    ],
)
@pytest.mark.filterwarnings("ignore:invalid escape sequence")
def test_parse_bytes(text, buffer) -> None:
    assert parse_bytes(text) == buffer


OBJECTS = [
    Versions.from_bytes(b"CFA533: h1.4, u1v2"),
    LcdMemory.from_bytes(b"\xff\x00\x01\x02\x03\x04\x05\x06\x07"),
    DowDeviceInformation.from_bytes(b"\xff\x00\x01\x02\x03\x04\x05\x06\x07"),
    DowTransactionResult.from_bytes(b"\xff\x01\x02\x03\x04\x05\x06\x07\xff"),
    KeypadPolled.from_bytes(bytes([KP_UP, KP_UP, KP_UP])),
    CFA533Status(
        temperature_sensors_enabled={1, 2},
        key_states=KeyStates(
            up=KeyState(
                keypress=KP_UP, pressed=False, pressed_since=True, released_since=True
            ),
            enter=KeyState(
                keypress=KP_ENTER,
                pressed=False,
                pressed_since=True,
                released_since=True,
            ),
            exit=KeyState(
                keypress=KP_EXIT, pressed=False, pressed_since=True, released_since=True
            ),
            left=KeyState(
                keypress=KP_LEFT, pressed=False, pressed_since=True, released_since=True
            ),
            right=KeyState(
                keypress=KP_RIGHT,
                pressed=False,
                pressed_since=True,
                released_since=True,
            ),
            down=KeyState(
                keypress=KP_DOWN, pressed=False, pressed_since=True, released_since=True
            ),
        ),
        atx_power_switch_functionality_settings=AtxPowerSwitchFunctionalitySettings(
            functions={AtxPowerSwitchFunction.KEYPAD_RESET},
            auto_polarity=True,
            reset_invert=False,
            power_invert=False,
            power_pulse_length_seconds=1.0,
        ),
        watchdog_counter=0,
        contrast=0.5,
        keypad_brightness=0.5,
        atx_sense_on_floppy=False,
        cfa633_contrast=0.5,
        lcd_brightness=0.5,
    ),
    GpioRead.from_bytes(bytes([0xFF, 0b0111, 0x11, 0b1101])),
    b"\01",
]


@pytest.mark.parametrize("obj", OBJECTS)
def test_repr(obj: Any, snapshot) -> None:
    assert repr(obj) == snapshot


@pytest.mark.parametrize("obj", OBJECTS)
def test_json(obj: Any, snapshot) -> None:
    assert json.dumps(as_json(obj)) == snapshot
