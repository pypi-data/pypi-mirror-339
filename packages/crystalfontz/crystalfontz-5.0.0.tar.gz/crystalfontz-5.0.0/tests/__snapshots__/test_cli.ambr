# serializer version: 1
# name: test_json[\x01]
  '"AQ=="'
# ---
# name: test_json[obj0]
  '{"model": "CFA533", "hardware_rev": "h1.4", "firmware_rev": "u1v2"}'
# ---
# name: test_json[obj1]
  '{"address": 255, "data": "AAECAwQFBgc="}'
# ---
# name: test_json[obj2]
  '{"index": 255, "rom_id": 283686952306183}'
# ---
# name: test_json[obj3]
  '{"index": 255, "data": "AQIDBAUGBw==", "crc": 255}'
# ---
# name: test_json[obj4]
  '{"states": {"up": {"keypress": 1, "pressed": true, "pressed_since": true, "released_since": true}, "enter": {"keypress": 2, "pressed": false, "pressed_since": false, "released_since": false}, "exit": {"keypress": 4, "pressed": false, "pressed_since": false, "released_since": false}, "left": {"keypress": 8, "pressed": false, "pressed_since": false, "released_since": false}, "right": {"keypress": 16, "pressed": false, "pressed_since": false, "released_since": false}, "down": {"keypress": 32, "pressed": false, "pressed_since": false, "released_since": false}}}'
# ---
# name: test_json[obj5]
  '{"temperature_sensors_enabled": [1, 2], "key_states": {"up": {"keypress": 1, "pressed": false, "pressed_since": true, "released_since": true}, "enter": {"keypress": 2, "pressed": false, "pressed_since": true, "released_since": true}, "exit": {"keypress": 4, "pressed": false, "pressed_since": true, "released_since": true}, "left": {"keypress": 8, "pressed": false, "pressed_since": true, "released_since": true}, "right": {"keypress": 16, "pressed": false, "pressed_since": true, "released_since": true}, "down": {"keypress": 32, "pressed": false, "pressed_since": true, "released_since": true}}, "atx_power_switch_functionality_settings": {"functions": [32], "auto_polarity": true, "reset_invert": false, "power_invert": false, "power_pulse_length_seconds": 1.0}, "watchdog_counter": 0, "contrast": 0.5, "keypad_brightness": 0.5, "atx_sense_on_floppy": false, "cfa633_contrast": 0.5, "lcd_brightness": 0.5}'
# ---
# name: test_json[obj6]
  '{"index": 255, "state": {"state": true, "falling": true, "rising": true}, "requested_level": 17, "settings": {"function": 8, "mode": 5, "up": "SLOW_STRONG", "down": "SLOW_STRONG"}}'
# ---
# name: test_repr[\x01]
  "b'\\x01'"
# ---
# name: test_repr[obj0]
  'CFA533: h1.4, u1v2'
# ---
# name: test_repr[obj1]
  '0xFF: \\x00\\x01\\x02\\x03\\x04\\x05\\x06\\x07'
# ---
# name: test_repr[obj2]
  '0xFF: 1020304050607'
# ---
# name: test_repr[obj3]
  '''
  Transaction Result for Device 255:
    Data: \x01\x02\x03\x04\x05\x06\x07
    CRC: 0xFF
  '''
# ---
# name: test_repr[obj4]
  '''
  Keypad States:
    up: keypress=KP_UP, pressed=yes, pressed_since=yes, released_since=yes
    enter: keypress=KP_ENTER, pressed=no, pressed_since=no, released_since=no
    exit: keypress=KP_EXIT, pressed=no, pressed_since=no, released_since=no
    left: keypress=KP_LEFT, pressed=no, pressed_since=no, released_since=no
    right: keypress=KP_RIGHT, pressed=no, pressed_since=no, released_since=no
    down: keypress=KP_DOWN, pressed=no, pressed_since=no, released_since=no
  '''
# ---
# name: test_repr[obj5]
  '''
  CFA533 Status:
  --------------
  Temperature sensors enabled: 1, 2
  Key states:
    up: keypress=KP_UP, pressed=no, pressed_since=yes, released_since=yes
    enter: keypress=KP_ENTER, pressed=no, pressed_since=yes, released_since=yes
    exit: keypress=KP_EXIT, pressed=no, pressed_since=yes, released_since=yes
    left: keypress=KP_LEFT, pressed=no, pressed_since=yes, released_since=yes
    right: keypress=KP_RIGHT, pressed=no, pressed_since=yes, released_since=yes
    down: keypress=KP_DOWN, pressed=no, pressed_since=yes, released_since=yes
  ATX Power Switch Functionality Settings:
    Functions enabled: KEYPAD_RESET
    Auto-Polarity Enabled: yes
    Reset Inverted: no
    Power Inverted: no
    Power Pulse Length (seconds): 1.0
  Watchdog Counter: 0
  Contrast: 0.5
  Contrast (CFA633 Compatible): 0.5
  Backlight:
    Keypad Brightness: 0.5
    LCD Brightness: 0.5
  '''
# ---
# name: test_repr[obj6]
  '''
  GPIO pin 255:
    Requested Level: 17
    Settings:
      Function: 8
      Drive Mode: SLOW_STRONG, SLOW_STRONG
  '''
# ---
